### PR TITLE
fix: address a data race in flinkapplication reconciler 

### PR DESCRIPTION
**Pull request by aburlutskiy**

### CODE REVIEW

File: pkg/controller/flinkapplication/controller.go [hunk 1]
```diff
@@ -82,10 +83,52 @@ func (r *ReconcileFlinkApplication) getReconcileResultForError(err error) reconc
...
+	mx: sync.RWMutex{},
+	m:  map[string]*sync.Mutex{},
+}
+
 func (r *ReconcileFlinkApplication) Reconcile(request reconcile.Request) (reconcile.Result, error) {
 	ctx := context.Background()
 	ctx = contextutils.WithNamespace(ctx, request.Namespace)
 	ctx = contextutils.WithAppName(ctx, request.Name)
+
+	key := string(request.Namespace) + "." + string(request.Name)
```
Comment (mdreizin) [inline pkg/controller/flinkapplication/controller.go]: Please use `fmt.Sprintf`:

```suggestion
	key := fmt.Sprintf("%s.%s", request.Namespace, request.Name)
```

File: pkg/controller/flinkapplication/controller.go [hunk 2]
```diff
@@ -82,10 +83,52 @@ func (r *ReconcileFlinkApplication) getReconcileResultForError(err error) reconc
...
+	logger.Debugf(ctx, "Trying to get a Mutex for a resource: %v", key)
+	lock := locks.GetOrLoad(ctx, key, &sync.Mutex{})
```
Comment (mdreizin) [inline pkg/controller/flinkapplication/controller.go]: Could you please move `locks` into `ReconcileFlinkApplication` struct?

Comment (mdreizin) [reply pkg/controller/flinkapplication/controller.go]: Because right now it is global, but it should be per an instance.

**Review by mdreizin — CHANGES_REQUESTED** [review]

File: pkg/controller/flinkapplication/controller.go (hunk 2 above)
Comment (aburlutskiy) [reply pkg/controller/flinkapplication/controller.go]: Ok. I will do that.

**Comment by aburlutskiy** [comment]
@mdreizin thank you for comments. I addressed all requested changes, please take a look.

File: pkg/controller/flinkapplication/controller.go [hunk 3]
```diff
@@ -36,6 +38,7 @@ type ReconcileFlinkApplication struct {
 	cache             cache.Cache
 	metrics           *reconcilerMetrics
 	flinkStateMachine FlinkHandlerInterface
+	locks             *SyncMap
```
Comment (mdreizin) [inline pkg/controller/flinkapplication/controller.go]: You can may just use [`m  sync.Map`](https://pkg.go.dev/sync#Map) to synchronize state:

```suggestion
```

File: pkg/controller/flinkapplication/controller.go [hunk 4]
```diff
@@ -82,10 +85,47 @@ func (r *ReconcileFlinkApplication) getReconcileResultForError(err error) reconc
...
+		defer c.mx.Unlock()
+		if val, found := c.m[key]; found {
+			return val
+		} else {
+			logger.Debugf(ctx, "There is no mutex for key: %v", key)
+			c.m[key] = lock
+			return lock
+		}
+	}
+}
```
Comment (mdreizin) [inline pkg/controller/flinkapplication/controller.go]: Please see my comment above.

```suggestion
```

**Review by mdreizin — CHANGES_REQUESTED** [review]

File: pkg/controller/flinkapplication/controller.go (hunk 3 above)
Comment (aburlutskiy) [reply pkg/controller/flinkapplication/controller.go]: Yes, that struct exists but I was inspired by some resources which said that sync.Map isn't a solution for all cases. And they supposed that map+RWMutex is more efficient then sync.Map where there is no high-concurrency read and write operations. 

https://habrastorage.org/r/w1560/web/fc8/179/905/fc81799050554f20a7b28314b68afe47.png
https://wrfly.kfd.me/posts/rwmutex-and-sync.map/

Comment (mdreizin) [reply pkg/controller/flinkapplication/controller.go]: That was in 2018.

Comment (mdreizin) [reply pkg/controller/flinkapplication/controller.go]: Go team has fixed many of them, but there are a couple of rare cases when built-in implementation is slow.

Please see https://github.com/golang/go/issues/28938

**Comment by aburlutskiy** [comment]
I switched to use sync.Map structure instead of using own structure. Please @mdreizin take a look again

File: pkg/controller/flinkapplication/controller.go [hunk 5]
```diff
@@ -82,10 +85,29 @@ func (r *ReconcileFlinkApplication) getReconcileResultForError(err error) reconc
 	}
 }
 
+type SyncMap struct {
+	mx sync.RWMutex
+	m  map[string]*sync.Mutex
+}
```
Comment (mdreizin) [inline pkg/controller/flinkapplication/controller.go]:
```suggestion
```

**Review by mdreizin — CHANGES_REQUESTED** [review]

**Review by mdreizin — APPROVED** [review]

**Pull request merged**